R basics introduction (#60)
arcus/education_modules: r_basics_introduction/r_basics_introduction.md

**First**, we need to create a small container in the cloud for you to work in just using your web browser.  **Click "Launch binder" below.**  It might take a while (5 minutes) to create, depending on how recently it was created (when it's being used more, it's quicker!).  We're looking for a faster way to get you off and running in RStudio without downloads and without creating accounts, but for now this is a great, free way for us to get you working with no extra work on your part.

  <a href = "https://mybinder.org/v2/gh/arcus/education_r_environment/main?urlpath=rstudio" target = "_blank"><img src="https://mybinder.org/static/images/badge_logo.svg"></a> **← Click the "launch binder" button!**

<div class = "hint" style = "align-items: center; display: flex;">

<div style = "margin: 1rem; max-width: 45%; float:left;"> If you're the first person to fire up this environment in a while, you might see this loading screen for several minutes.  Be patient!</div>
<div style = "margin: 1rem auto; max-width: 45%; float:left;"> ![Binder loading screen](media/binder_loading.gif)<!--
style = "border: 1px solid rgb(var(--color-highlight));"-->
</div>
</div>

> **Then**, once you have access to RStudio and you see something like the image below, you'll need to open the sample data for this course.  In the file area to the lower right, you'll see, among multiple choices, the folder called "r\_basics\_visualize\_data".  That's the code for this module!

Author: Well I have an animated .gif that shows how to connect to the materials from your own computer, so it's treated a bit differently in the "server" versus "your computer" sections.  I think it makes sense to include it here.

![RStudio as shown in the cloud platform Binder](media/binder_rstudio.png)<!--
style = "border: 1px solid rgb(var(--color-highlight)); max-width: 800px;"-->

<h3>Option 2: Work on Your Computer</h3>

> If you have [R](https://www.r-project.org/) and [RStudio](https://www.rstudio.com/products/rstudio/download/#download) installed already on your local computer, you might be interested in simply downloading our sample code to your computer. Here's how.  Note -- if you've already done this step in another module, you might have the material for this module already!

Author: I also have other (different) comments from Elizabeth on the transform module, so I will do some side-by-side diff once all of the other issues are resolved and make sure I get the best of all worlds.  Holding off on this for now.

> If you have [R](https://www.r-project.org/) and [RStudio](https://www.rstudio.com/products/rstudio/download/#download) installed already on your local computer, you might be interested in simply downloading our sample code to your computer. Here's how.  Note -- if you've already done this step in another module, you might have the material for this module already!

Reviewer: sounds good to me!

> If you have [R](https://www.r-project.org/) and [RStudio](https://www.rstudio.com/products/rstudio/download/#download) installed already on your local computer, you might be interested in simply downloading our sample code to your computer. Here's how.  Note -- if you've already done this step in another module, you might have the material for this module already!

Author: OK I have parallelized the language!

### Future You

Reproducibility doesn't only help people **outside** a study understand how things were done.  They also help the initial conduct of research as well.  Consider the following 3 statements and ask yourself if they sound familiar.

- Can we redo the analysis with this month's data?
- Why do the data in Table 1 not seem to agree with Figure 2?
- Why did I decide to omit these six samples from my analysis?

**Your closest collaborator is you from 6 months ago...**

...  but they don't answer your emails!

Additionally, we've already seen how irreproducibility can cause real harm to human subjects, something we all want to avoid.  

In short, reproducible research can reduce harm for subjects, science, and scientists.  There are tools to help you conduct research data analysis in a way that helps your future collaborators, including future you, understand and extend your work.  A great tool for reproducible research is the use of R Markdown to record your research steps.

> 

Author: I think it's helpful reinforcement, and it tees up why R Markdown matters.  The "reproducibility" module doesn't get down to use cases in this level of specificity, so I think this wouldn't be obvious there, either.  I'll work on some transitional text to smooth out the seams, though!

> 

Reviewer: That's fair, my concern is just "scope creep" -- the stated learning objectives of this module are all about getting someone familiar and comfortable with R, RStudio, and R Markdown, not convincing them of the importance of why things like literate statistical programming are important. If we want these modules to be truly "just in time" it might make more sense to have motivational content like this in a separate module like "Why X matters for the unconvinced". 

I also think this content could be super valuable for all sorts of folks, including those who only want to learn Python, e.g. 
While we know that folks in our pilot will all get this content, once the content is treated as intended (broadly disseminated a la carte usage as needed by a variety of research folks), this content will only be reaching people who want to learn R/RStudio/R Markdown and are beginning their education at the novice level. 

That being said, of course the editorial decision lies with you, and I don't think it would be terrible to leave it here!  -- just wanted to make my position a bit clearer. :)

To **knit** a document, click the button that says "Knit" next to a blue ball of yarn on the top of the Editor pane.  Then in a few moments, a rendered document will be shown:

![RStudio window with the knit button indicated](media/knit_button.png)<!-- style = "max-width:600px;"-->
![Result of knitting, the start of an html document with headers and code](media/rendered_markdown.png)<!-- style = "max-width:600px;"-->

### R Markdown in Detail

Let's look at this R Markdown document in a little more detail.  Below is an R Markdown document and the resulting output in HTML that was created when this R Markdown document was knitted.

Below these images, we'll explain what each part of the R Markdown includes in a bit more detail.

<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 50%; float:left;">
![25 lines of R Markdown](media/r_markdown_detail.png)
</div>
<div style = "margin: 1rem; max-width: 50%; float:left;">

> ![A brief report with headers, some statistics, and a graph](media/rendered_detail.png)

Author: This might be too wordy but I replaced with the following.  LMK what you think!

<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 50%; float:left;">
![R Markdown code.  It begins with a short header enclosed by lines containing three dashes. Then text with a few special markdown characters alternates with code chunks.  Code chunks are preceded by a line with three backticks and the letter r in curly braces, and are followed by a line with three backticks. ](media/r_markdown_detail.png)
</div>
<div style = "margin: 1rem; max-width: 50%; float:left;">

![A brief report that is created when the R Markdown code is rendered.  The R Markdown header appears in the form of a large title.  The R Markdown text appears as appropriately rendered text with bold, italics, and other decorations.  The code chunks are represented by code and code output, such as a graph.](media/rendered_detail.png)
</div>
</div>

> ![A brief report with headers, some statistics, and a graph](media/rendered_detail.png)

Reviewer: made some suggestions

> ![A brief report with headers, some statistics, and a graph](media/rendered_detail.png)

Author: included!

### The Tidyverse

<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 75%; float:left;">
To import our CSV data, we need some additional data analysis tools.  In this course, we will be leveraging the **tidyverse**.

The tidyverse is a set of tools that has become the de facto standard for doing data science with R.  It relies on the use of "tidy" data and "tidy" data analysis techniques.
</div>
<div style = "margin: 1rem; max-width: 25%; float:left;">
![Tidyverse logo](media/tidyverse_logo.png)<!-- style = "max-width:300px;" -->
</div>
</div>

The basic tenets of "tidy" data analysis include:

* Data should be organized in a consistent, standardized way. Each row is an observation, and each column is a variable. This is a very common way to organize data in a spreadsheet and might sound familiar from how you may already organize data in tools like Excel.
* Programming code that acts on the data should be consistent, concise, and sound like human language as much as possible.

> * Each data analysis can be broken down into a series of atomic steps, such as "select this column" or "arrange the data by the values in that column". An arbitrarily complex data analysis can be broken down as a pipeline of atomic steps.

Author: agree this belongs in tidy data module.  Removed.

You may be familiar with functions from math class or perhaps also from formulas in Excel. A function takes an input, say an 'x' value, and returns an output, say a 'y' value. Functions in computer programming also take inputs and return outputs. But the inputs and outputs here are the objects that exist within the context of a programming language.

For `read_csv`, the input is the file name of a CSV file; and the output is a data frame with the contents of the file.

In the snippet of code that reads in a csv to create a data frame, we have a **function** with an **argument** that creates an **object**.

The input that goes into a function is called an **argument**. The argument to a function gets put in parentheses.

A function can have zero, one, or many arguments. If there is more than one argument, we use commas to separate them. We'll see examples of that later.

The output of the read_csv function is a data frame object. You will usually want to *capture* that data frame inside of a named object so you can use it as the input for other functions, for example to summarize or visualize your data.

> ![`The code is annotated. "data_frame" is labeled "object", the leftward arrow <- is labeled "assignment operator", "read_csv" is labeled "function", and "file_name" is labeled "argument".](media/function_breakdown.png)<!-- style = "max-width:500px;" -->

Author: agree, fixed

Let's put these pieces together to load our covid data set.

Consider this line of code:

``` r
covid_testing <- read_csv("covid_testing.csv")
```
<div class = "warning">
We could speak aloud the line of code above saying "covid underscore testing gets read underscore csv of covid underscore testing dot csv", or we could gloss over the underscores and just say "covid testing" and "read csv".  When we talk with others about R, this is often what we do, so just be aware that you sometimes have to intuit underscores when someone is reading aloud or talking through a code issue.
</div>

<h3>Quotes</h3>

You might notice that one of our "covid testing" objects is put in quotes and the other one isn't.

To be honest, quotes in R can be super confusing. A couple of rules are:

* Names of objects such as data frames don't get quotes.
* Literal file names are always put in quotes.
* When you're referring to a string value, like "abc123" or "jones", you have to use quotes.
* Names of packages, when you're installing especially, use quotes
* But in `library`, putting the package name in quotes is optional!

> Sometimes, either quoting or not quoting works. This is one of the things where you will get a good sense but not until after some trial and error.

Author: Yep that makes much more sense, I've restructured things.  We should maybe back-propagate some of these structural changes into the workshop format, too!

[[ ]] Functions use one or more inputs called arguments
[[ ]] You must install a package before using it in each new session of R
[[X]] You must load a package before using it in each new session of R
[[X]] The `install.packages()` function requires putting the package name in quotes
[[?]] Hint: multiple correct answers!

<div class = "answer">

<details><summary>Click to see an explanation of the answer.</summary>

Functions can have zero, one, or many inputs.  An example of a function taking zero inputs is `getwd()`, which tells you what R considers its current "working directory".

You only need to install a package using `install.packages()` **once** per computer (unless the package has a new update that you want to install).  You do, however, have to load a package using `library()` before its functions are available to you.  That means loading it each time you start a new session of R.  **When you knit a document, that counts as a separate session**, so make sure and put a `library()` in place for each package you use in your code in each R Markdown document you create.  Oh, and yep, for `install.packages()` you do indeed have to put quotes around the name of the package.

> </div>

Author: at times I just throw a <lia-keep> around html tags to make sure they don't render wrong.  Pretty sure I got this bug squashed.

</div>

### Hands-On: Import Data

Please go ahead and open the "r\_basics\_visualize\_data" directory in whichever RStudio environment you're working in. You can open `introduction_exercises.Rmd`.  Instructions for this exercise are in the text of the R Markdown document.

![`The first few lines of the introduction_exercises.Rmd file`](media/introduction_exercises.png)<!-- style = "max-width:700px;" -->

Read through the instructions and complete the assigned tasks:

* import data from a .csv file
* use the RStudio data viewer to answer some questions

### Solutions

You can open the `introduction_solutions.Rmd` file to see the answers.

> To import data, you run the cell indicated.  Click on the image below to show a short clip of what this looks like.

Author: Here's the html with the alt text:

```
  <img src="https://github.com/arcus/education_modules/blob/r_basics_introduction/r_basics_introduction/media/run_import_chunk.png?raw=true" height="540" width="878" alt="The file introduction_exercises.Rmd is open.  The user scrolls to line 17, which is the first line of a code chunk.  In the upper right corner of this code chunk, there is a green play button which is clicked." data-alt="https://github.com/arcus/education_modules/blob/r_basics_introduction/r_basics_introduction/media/run_import_chunk.gif?raw=true" style = "border: 1px solid rgb(var(--color-highlight));">

```

<div style="display:none">@gifPreload</div>

<figure>
  <img src="https://github.com/arcus/education_modules/blob/r_basics_introduction/r_basics_introduction/media/run_import_chunk.png?raw=true" height="540" width="878" alt="RStudio can create a new project that gets its contents from a git repository." data-alt="https://github.com/arcus/education_modules/blob/r_basics_introduction/r_basics_introduction/media/run_import_chunk.gif?raw=true" style = "border: 1px solid rgb(var(--color-highlight));">

<figcaption style = "font-size: 1em;">Click on the image to play the demo of running the chunk that imports data.</figcaption>
</figure>

> To answer the questions using the file viewer, click on the name of a data frame (like `covid_testing`) to open a view into the data in a new tab.  Click on the image below to show a short clip of what this looks like.

Author: New alt text: alt="The mouse moves to the Environment pane of RStudio and the named object covid_testing is clicked, and a view into that data opens in the upper left pane of RStudio."

<figcaption style = "font-size: 1em;">Click on the image to play the demo of opening a viewer of a data frame.</figcaption>
</figure>

Here are the answers to the questions posed in `introduction_exercises.Rmd`:

1. How many rows are in the data frame? How many columns?

There are 15,524 rows and 17 columns. This information can be seen at the bottom of the Data Viewer, and also in the Environment pane. Note that in R parlance, rows are observations, and columns are variables.

2. Go ahead and try to edit one of the values in this viewer. You will find that you can't. It would have been easy for the RStudio programmers to allow editing of specific values, but they decided not to add that feature. Why do you think this was designed that way?

This would encourage point-and-click workflows that are not reproducible.

> 3. Each row in *covid_testing* represents a single COVID19 test. The value in the *pan_day* column indicates the day (after the start of the pandemic) on which the test sample was taken. On which "pandemic day" was the first test sample taken? The last?

Reviewer: weirdness with italics and underscores and things ... perhaps an escape is needed?

Author: whoops yep you are right!

If you sort the table by `pan_day` and then look at the very top and the very bottom, you will find out the first "pandemic day" was 4, and the last one was 107. You can sort the data in the Data Viewer by clicking on the column header.

Important: this doesn't actually change the data in the `covid_testing` data frame!

4. How many tests overall were positive? How many positive tests were there in the first 30 days of the pandemic? (Hint: look at the Filter button)

The `result` column holds that information. We can apply a filter to that column to only show us rows where the value in the `result` column is "positive". Click on Filter and enter "positive" in the text field that appears below the `result` header. You will now see that the bottom line of the data viewer reads "Showing 1 to [some number] of 865 entries, 17 total columns (filtered from 15,524 total entries)". So there were 865 positive tests, out of 15,524 total.

## Recap

Let's recap what we have covered in this session.

> ![R logo](media/r_logo.png)<!-- style="max-width:300px; margin:2em auto;" --> ![RStudio logo](media/rstudio_logo.png)<!-- style="max-width:300px; margin:2em auto;" -->![R Markdown logo](media/r__markdown_logo.png)<!-- style="max-width:300px; margin:2em auto;" -->

Reviewer: these images are rendering strangely (and RMarkdown isn't showing up at all)

Author: I can't get things to show up small and side by side, it's super annoying.  I'm just taking these out, as they don't really add anything.

> ![R logo](media/r_logo.png)<!-- style="max-width:300px; margin:2em auto;" --> ![RStudio logo](media/rstudio_logo.png)<!-- style="max-width:300px; margin:2em auto;" -->![R Markdown logo](media/r__markdown_logo.png)<!-- style="max-width:300px; margin:2em auto;" -->

Reviewer: ```suggestion
![""](media/r_logo.png)<!-- style="max-width:300px; margin:2em auto;" --> ![""](media/rstudio_logo.png)<!-- style="max-width:300px; margin:2em auto;" -->![""](media/r__markdown_logo.png)<!-- style="max-width:300px; margin:2em auto;" -->
```

We started by defining and differentiating **R** (the programming language) from **RStudio** (the integrated development environment -- IDE -- or editor) and **R Markdown** (the document format we use for reproducible data analysis). Then we explored RStudio and R Markdown more closely.

<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 30%; float:left;">

> ![The tidyverse package is shown alongside a brown box representing a package](media/packages.png)<!-- style = "max-width: 400px;" -->

Reviewer: ```suggestion
![""](media/packages.png)<!-- style = "max-width: 400px;" -->
```

</div>
<div style = "margin: 1rem; max-width: 70%; float:left;">
Packages extend the functionality of R. Install with `install.packages()` and load with `library()`.
</div>
</div>
<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 30%; float:left;">

> ![`The read_csv function is shown with its parts labeled`](media/function_breakdown.png)<!-- style = "max-width: 400px;" -->

Reviewer: ```suggestion
![""](media/function_breakdown.png)<!-- style = "max-width: 400px;" -->
```

</div>
<div style = "margin: 1rem; max-width: 70%; float:left;">
Functions do stuff. They accept Arguments as input and return an Output. Capture an output in an Object using the assignment operator (` <- `).
</div>
</div>
<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 30%; float:left;">

> ![csv file is transformed into a data frame](media/csv_import.png)<!-- style = "max-width: 400px;" -->

Reviewer: ```suggestion
![""](media/csv_import.png)<!-- style = "max-width: 400px;" -->
```

Importing Data is the first step of data analysis. Use `read_csv()` from the tidyverse package to import data stored in a CSV file.
</div>
</div>

## Additional Resources

<h3>Cheat Sheets</h3>

RStudio's Data Import cheat sheet helps with importing data from file types similar to CSV but with other delimiters.  Cheat sheets for the RStudio IDE and Markdown can reinforce what you've learned so far in this module.

To find official cheat sheets, go to the Help menu and choose "Cheat Sheets".  You can look around the short list provided.  You can also "Browse", and in the RStudio website you'll find other cheat sheets as well, including one for

![RStudio help menu, with Cheat Sheets selected and the submenu option "Browse Cheat Sheets" selected](media/cheat_sheet_browse.png)<!-- style = "max-width:700px;" -->

### More Data Types

<h3>File Formats</h3>

Not all data is in CSV files, and fortunately R supports a huge number of other file formats, and here are some of the packages that help import or export them:

* Readxl: Microsoft Excel
* Haven: SPSS, Stata, and SAS files
* Googlesheets: Google Sheets
* Rvest: web scraping
* Jsonlite: JSON data and many APIs on the web

> ![Hexagonal logos of the five packages listed above](media/file_types.png)<!-- style = "max-width:600px;" -->

Reviewer: ```suggestion
![""](media/file_types.png)<!-- style = "max-width:600px;" -->
```

<h3>Databases</h3>

You can also connect to a large number of **databases** directly in R to pull your data. For example, you can connect to SQL databases including MySQL and Oracle, as well as using APIs (application programming interfaces) for data collection applications like REDCap and websites with API access like PubMed or the New York Times.

> ![Logos of many different database solutions](media/databases.png)

Reviewer: ```suggestion
![""](media/databases.png)
```

### Additional Features

<h3>Output Document Options</h3>

<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 50%; float:left;">
In addition to HTML, R Markdown documents can be "knitted" into a number of additional formats including PDF, Microsoft Word, PowerPoint, and even interactive dashboards.
</div>
<div style = "margin: 1rem; max-width: 50%; float:left;">

> ![R Markdown logo surrounded by the logos of output document options, including HTML5, .pdf, and others.](media/document_output.png)<!-- style = "max-width: 400px;" -->

Reviewer: ```suggestion
![""](media/document_output.png)<!-- style = "max-width: 400px;" -->
```

</div>
</div>

<h3>Language Support</h3>

Finally, in addition to R, R Markdown supports many other programming languages, including Python.

So you can mix code chunks written in R with code chunks written in, say, Python:

<div style = "align-items: center; display: flex;">

<div style = "margin: 1rem; max-width: 50%; float:left;">

> ![R logo with cartoon snake](media/r_python.png)<!-- style = "max-width: 400px;" -->

Reviewer: ```suggestion
![""](media/r_python.png)<!-- style = "max-width: 400px;" -->
```

</div>
<div style = "margin: 1rem; max-width: 50%; float:left;">

<lia-keep>
<pre>
```{python}
import pandas
covid_testing.info()
```
</pre>
</lia-keep>
</div>
</div>

> 

Reviewer: most of this content does not belong in additional resources -- this should go in another section. 

Also do you have any other resources we can link to currently in the additional resources? (internally + externally)

Author: Agree, I've moved this section up, and slightly beefed up the "cheat sheets" section in Additional Resources.

## Feedback

In the beginning, we stated some goals.

**Learning Objectives:**

@learning_objectives

We ask you to fill out a brief (5 minutes or less) survey to let us know:

* If we achieved the learning objectives
* If the module difficulty was appropriate
* If we gave you the experience you expected

> We gather this information in order to iteratively improve our work.  Thank you in advance for filling out [our brief survey](https://redcap.chop.edu/surveys/?s=KHTXCXJJ93&module_name=%22R+Basics+Visualize+Data%22)!

Reviewer: ```suggestion
We gather this information in order to iteratively improve our work.  Thank you in advance for filling out [our brief survey](https://redcap.chop.edu/surveys/?s=KHTXCXJJ93&module_name=%22R+Basics+Introduction%22)!
```

		image[index].src = gif[index];
	});

	// Change the image to .gif when clicked and vice versa.
	$('figure').on('click', function() {

		var $this   = $(this),
				$index  = $this.index(),

				$img    = $this.children('img'),
				$imgSrc = $img.attr('src'),
				$imgAlt = $img.attr('data-alt'),
				$imgExt = $imgAlt.split('.');

		if($imgExt[1] === 'gif') {
			$img.attr('src', $img.data('alt')).attr('data-alt', $imgSrc);
		} else {
			$img.attr('src', $imgAlt).attr('data-alt', $img.data('alt'));
		}

		// Add play class to help with the styling.
		$this.toggleClass('play');

	});

})(jQuery);
</script>
@end

link:  https://chop-dbhi-arcus-education-website-assets.s3.amazonaws.com/css/styles.css
script: https://kit.fontawesome.com/83b2343bd4.js
-->
# R Basics: Introduction

<div class = "overview">

## Overview

@comment

**Is this module right for me?** @long_description

**Estimated time to completion:** @estimated_time

**Pre-requisites**

No prior experience of using R, RStudio, or R Markdown is required for this course.   

This course is designed for brand new beginners with zero or minimal experience working with R.

**Learning Objectives**

@learning_objectives

</div>

> Material for this module was adapted, with permission, from [Stephan Kadauke's R for Clinical Data workshop materials](https://skadauke.github.io/intro-to-r-for-clinicians-chop/).  We owe special thanks to Dr. Kadauke as well as the [R User Group](https://redcap.chop.edu/surveys/?s=NPY49R9ARF) at Children's Hospital of Philadelphia for their generosity in sharing these materials.

Reviewer: ```suggestion
Material for this module was adapted, with permission, from [Stephan Kadauke's R for Clinical Data workshop materials](https://skadauke.github.io/intro-to-r-for-clinicians-chop/).  We owe special thanks to Dr. Kadauke as well as the R User Group at Children's Hospital of Philadelphia for their generosity in sharing these materials.
```

The second code chunk demonstrates that (1) in addition to text output, you can also make R create graphical plots, and (2) you can tweak the behavior of a code chunk – note that this code chunk has a chunk option of `echo=FALSE` within the curly braces. This tells R that you don't want it to repeat, or "echo", the code in the rendered document.  Notice that while the code itself is suppressed in the rendered output document, the results of the code (that is, the graph), does appear.

Now let's practice creating an R Markdown document, running a code chunk, and knitting.

> ### Practicing R Markdown

Reviewer: ```suggestion
### Hands-On: Practicing R Markdown
```
to match w/ import data activity

![`The data analysis pipeline, which begins with Import, then Tidy, then three actions that form a circle that repeat: Visualize, Model, and Transform.  The pipeline ends with "Communicate".`](media/pipeline_import.png)<!-- style = "max-width:500px;"-->

<figcaption style = "font-size: 0.8em;"> Image taken from *R for Data Science* (https://r4ds.had.co.nz/introduction.html)</figcaption>

The first step in the data analysis pipeline is to bring in, or **import**, the data to your environment, which for the purposes of this module is R.

In this module, we'll work with a fabricated (completely made up) data set consisting of COVID-19 laboratory test results from a hypothetical infectious disease diagnostics lab. This data is stored as a **CSV file**.

So what's a .csv file? CSV stands for **comma-separated values**.  When we say this aloud, we just say each letter: C - S - V.  

A .csv file is a plain text file, which means you can open it in a text editor and look at it.  This is different than how you may have opened .csv files in the past, using Microsoft Excel or another spreadsheet program.  Let's take a look at what the raw data looks like when we open it in a text editor.

> ![when viewed as plain text, csv files are just values separated by commas within each "row", with each row returned on a new row. ](media/csv.png)<!-- style = "max-width:500px;" -->

Reviewer: ```suggestion
![when viewed as plain text, csv files are just values separated by commas within each "row", and each row returned on a new line. ](media/csv.png)<!-- style = "max-width:500px;" -->
```

I know this was my own suggestion last time... but reading it with fresh eyes made me realize it did not make as much sense as I thought it did lol

``` r
data_frame <- read_csv(file_name)
```
![""](media/csv_import.png)<!-- style = "max-width:500px;" -->

> You start with the name of the object you want to hold the data, then you type a leftward arrow symbol using the less than followed by a dash symbol, then "read_csv" and then the file name in parentheses.

Reviewer: ```suggestion
You start with the name of the object you want to hold the data, then you type a leftward arrow symbol using the less than symbol followed by a dash symbol, then "read_csv" and then the file name in parentheses.
```

> A function can have zero, one, or many arguments. If there is more than one argument, we use commas to separate them. We'll see examples of that later.

Reviewer: ```suggestion
A function can have zero, one, or many arguments. If there is more than one argument, we use commas to separate them.
```

We don't actually see any examples like that in this module, right? I think "later" is a remnant from the workshop when all the content was in one place.

[[?]] Hint: multiple correct answers!

<div class = "answer">
<details><summary>Click to see an explanation of the answer.</summary>

Functions can have zero, one, or many inputs.  An example of a function taking zero inputs is `getwd()`, which tells you what R considers its current "working directory".

You only need to install a package using `install.packages()` **once** per computer (unless the package has a new update that you want to install).  You do, however, have to load a package using `library()` before its functions are available to you.  That means loading it each time you start a new session of R.  **When you knit a document, that counts as a separate session**, so make sure and put a `library()` in place for each package you use in your code in each R Markdown document you create.  Oh, and yep, for `install.packages()` you do indeed have to put quotes around the name of the package.

</details>

</div>
</div>

> ### Hands-On: Import Data

Reviewer: Do we potentially need to include more instruction before throwing them into this section, or is it fine? I'm mostly thinking about the fact that most of these exercises don't have anything to do with importing, they moreso have to do with clicking around in RStudio, exploring data, and noticing stuff in the data viewer. 

I can't remember how much of that sort of stuff they already saw us demo in the workshop by the time we send them to do this.

Author: I think this is a good catch.  I'll at the very least add a screenshot here with good alt text.

Author: Upon rereading this I think I missed your point but stumbled upon a new one, ha!  I think we are gentle enough in the introduction to data importing and in the interest of time I'll leave this be.  But I did add a bit of enrichment in the "if you have to install tidyverse on your own computer" callout box since many learners may be issuing their very first line of code....

<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 25%; float:left;">
![""](media/packages.png)<!-- style = "max-width: 100%;" -->
</div>
<div style = "margin: 1rem; max-width: 70%; float:left;">
Packages extend the functionality of R. Install with `install.packages()` and load with `library()`.
</div>
</div>
<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 25%; float:left;">
![""](media/function_breakdown.png)<!-- style = "max-width: 100%;" -->
</div>
<div style = "margin: 1rem; max-width: 70%; float:left;">
Functions do stuff. They accept Arguments as input and return an Output. Capture an output in an Object using the assignment operator (` <- `).
</div>
</div>
<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 25%; float:left;">
![""](media/csv_import.png)<!-- style = "max-width: 100%;" -->
</div>
<div style = "margin: 1rem; max-width: 70%; float:left;">
Importing Data is the first step of data analysis. Use `read_csv()` from the tidyverse package to import data stored in a CSV file.
</div>
</div>

## Additional Resources

<h3>Cheat Sheets</h3>

> To find official cheat sheets, go to the Help menu in RStudio and choose "Cheat Sheets".  You can look around the short list provided, which includes great options for beginners such as "RStudio IDE Cheat Sheet", "R Markdown Cheat Sheet", and more.  You can also "Browse", and in the RStudio website you'll find other cheat sheets as well, including the Data Import cheat sheet.

Reviewer: ```suggestion
To find official cheat sheets, go to the Help menu in RStudio and choose "Cheat Sheets".  You can look around the short list provided, which includes great options for beginners such as "RStudio IDE Cheat Sheet", "R Markdown Cheat Sheet", and more.  You can also "Browse", and on the RStudio website you'll find other cheat sheets as well, including the Data Import cheat sheet.
```

RStudio's Data Import cheat sheet helps with importing data from file types similar to CSV but with other delimiters (for example, TSV, or tab-delimited values files).  Cheat sheets for the RStudio IDE and Markdown can reinforce what you've learned so far in this module and help you extend your capabilities.

![RStudio help menu, with Cheat Sheets selected and the submenu option "Browse Cheat Sheets" selected](media/cheat_sheet_browse.png)<!-- style = "max-width:700px;" -->

### R for Data Science

<div style = "align-items: center; display: flex;">
<div style = "margin: 1rem; max-width: 60%; float:left;">
*R for Data Science* is a free text that provides lots of helpful explanation and examples.  The [section on data import](https://r4ds.had.co.nz/data-import.html) goes into much more detail than we were able to do in this brief module.

> The entire text is available [in English](https://r4ds.had.co.nz/) and [in Spanish](https://es.r4ds.hadley.nz/).

Reviewer: ```suggestion
The entire text is available [in English](https://r4ds.had.co.nz/) and [in Spanish](https://es.r4ds.hadley.nz/).
There's also an [unofficial solutions guide](https://jrnold.github.io/r4ds-exercise-solutions/index.html) (only available in English) to allow you to check your work. 
```

*R for Data Science* is a free text that provides lots of helpful explanation and examples.  The [section on data import](https://r4ds.had.co.nz/data-import.html) goes into much more detail than we were able to do in this brief module.

The entire text is available [in English](https://r4ds.had.co.nz/) and [in Spanish](https://es.r4ds.hadley.nz/).
There's also an [unofficial solutions guide](https://jrnold.github.io/r4ds-exercise-solutions/index.html) (only available in English) to allow you to check your work. 
</div>
<div style = "margin: 1rem; max-width: 20%; float:left;">

> ![Book Cover](media/r4ds.png)<!-- style = "max-width: 90%" -->

Reviewer: ```suggestion
![""](media/r4ds.png)<!-- style = "max-width: 90%" -->
```

Reviewer: @pm0kjp just this one final little change and we should be good to merge it in! 😄
